### PR TITLE
Add Release Notes Guidance to CONTRIBUTING

### PR DESCRIPTION
**Pull request by pow2clk**
Add guidance for how release notes should be documented at the time
of the change going in as well as some suggestions for how to format
that documentation.

Contributes to #6697

### CODE REVIEW

**Comment by pow2clk** [comment]
There were some incidental and not-quite-so-incidental reformatting of the original content that went in as the first two commits. Only the last commit pertains directly to release notes. Reviewing the commits in sequence might make the task easier: https://github.com/microsoft/DirectXShaderCompiler/pull/6778/commits

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+Titles under 76 characters print nicely in unix terminals under `git log`.
+This is not a hard requirement, but is good guidance.
+
+Tags in titles allow for speedy categorization
+Title tags  are generally one word or acronym enclosed in square brackets.
+Limiting to one or two tags is ideal to keep titles short.
+Some examples of common tags are:
+
+- `[NFC]` - No Functional Change
+  - `[RFC]` - Request For Comments (often used for drafts to get feedback)
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:73]: Indentation is off here.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+If a change is meant for the next upcoming release,
+it should be added to the "Upcoming Release" section.
+When the next upcoming release is named,
+the title will be updated and the release note will be included in the appropriate release.
+
+When writing release note list entries:
+
+- Keep the description to a single sentence.
+- Links to PRs are not recommended.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:121]: I would go further to say no links to PRs, but links to issues are allowed

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+- Markdown links to bugs are fine if the issue is too complicated to completely explain in a single sentence.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:122]: Maybe even encourage links to issues? They can be super helpful and compact.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+- Remember to update release notes as the nature of the change alters or is removed.
+
+### Testing Pull Requests
+
+For a pull request to be merged, it will have to pass the automated set of regression tests run for each.
+Additional regression testing may be required for some changes depending on the area of the changes.
+The commiter is expected to recognize the need for and perform any additional testing prior to merging.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:129]: I think this is a lot unrealistic and would put a significant burden on external contributions. Code reviewers should analyze changes and be able to make this determination.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+In addition to existing tests, bug fixes and new features require additional testing be included in the pull requests that implement them.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:131]: I would put this at the top of the section on testing since it is the most important thing. Maybe even add something explicit like:

> All PRs that contribute functional or behavioral changes to the compiler must have additional tests that verify the new code behaves correctly. Changes that do not change behavior may be required to add testing if the change impacts areas with poor test coverage.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+- `[NFC]` - No Functional Change
+  - `[RFC]` - Request For Comments (often used for drafts to get feedback)
+  - `[Doc]` - Documentation change
+  - `[SPIRV]` - Changes related to SPIR-V
+  - `[HLSL2021]` - Changes related to HLSL 2021 features
+  - Other tags in use: `[Linux]`, `[mac]`, `[Win]`, `[PIX]`, etc...
+
+Tags aren't formalized or any specific limited set. If you're unsure of
+  a reasonable tag to use, just don't use any. If you want to invent a new
```
Comment (damyanp) [inline CONTRIBUTING.md:80]: unnecessary indentation

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+- Visible to the users
+- Significant changes to compiler behavior
+  - New features: Language, Hardware support, compiler options
+  - Important bug fixes
+  - Changes in default behavior
+
+When such a change is made, the release note should be included as part of that change.
+This is done in the docs/ReleaseNotes.md file.
+
+If the change is meant for a named release,
```
Comment (damyanp) [inline CONTRIBUTING.md:108]: can we wrap lines consistently please?  I find this confusing to read - it looks like a poem.  Maybe I've missed a literary reference?

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+When writing release note list entries:
+
+- Keep the description to a single sentence.
+- Links to PRs are not recommended.
+- Markdown links to bugs are fine if the issue is too complicated to completely explain in a single sentence.
+- Remember to update release notes as the nature of the change alters or is removed.
+
+### Testing Pull Requests
+
+For a pull request to be merged, it will have to pass the automated set of regression tests run for each.
```
Comment (damyanp) [inline CONTRIBUTING.md:127]: > For a pull request to be merged, it will have to pass the automated set of regression tests run for each.

For each....? Maybe something missing at the end of the sentence?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -28,65 +28,118 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+In addition to existing tests, bug fixes and new features require additional testing be included in the pull requests that implement them.
+For bug fixes, at least one added test should fail in the absence of your non-test code changes.
+Tests should include reasonable permutations of the target fix/change.
+Include baseline changes with your change as needed.
+
+For cases where any of the above testing requirements are not possible,
+please specify why in the pull request.
+
+### Merging Pull Requests
```
Comment (damyanp) [inline CONTRIBUTING.md:139]: Anything to say about squashing vs not-squashing here?

File: CONTRIBUTING.md (hunk 4 above)
Comment (pow2clk) [reply CONTRIBUTING.md:129]: I agree. That's more for insiders with that access. I'll try to make that clearer.

File: CONTRIBUTING.md (hunk 7 above)
Comment (pow2clk) [reply CONTRIBUTING.md:108]: I write MD files this way in order to make follow-up changes easier to review since trying to review a long line of MD for changes somewhere therein is painful. 

The MD processor eliminates all single line breaks, so you can read it cleanly if you use "rich diff":
image

File: CONTRIBUTING.md (hunk 8 above)
Comment (pow2clk) [reply CONTRIBUTING.md:127]: I'll admit it's a bit hard to parse. The intent was to refer to regression tests run for each PR. I'll take another pass at it.

File: CONTRIBUTING.md (hunk 9 above)
Comment (pow2clk) [reply CONTRIBUTING.md:139]: It's beyond what I set out to do here, but yeah, I can make mention of it.

File: CONTRIBUTING.md (hunk 7 above)
Comment (damyanp) [reply CONTRIBUTING.md:108]: My feedback here is that I find this style unusual and it makes reading it harder.

File: CONTRIBUTING.md (hunk 8 above)
Comment (pow2clk) [reply CONTRIBUTING.md]: I did the default thing in editing and split it into two sentences.

File: CONTRIBUTING.md (hunk 7 above)
Comment (pow2clk) [reply CONTRIBUTING.md]: I've relaxed my line length threshold and merged most of these lines while making those left look more like continuation lines.

File: CONTRIBUTING.md (hunk 4 above)
Comment (pow2clk) [reply CONTRIBUTING.md]: Done

File: CONTRIBUTING.md (hunk 9 above)
Comment (pow2clk) [reply CONTRIBUTING.md:137]: Done

File: CONTRIBUTING.md (hunk 5 above)
Comment (pow2clk) [reply CONTRIBUTING.md]: Moved. I tried to meld the new suggestions with the existing text.

**Review by pow2clk — COMMENTED** [review]
Responding to feedback comments

**Review by damyanp — APPROVED** [review]

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -28,65 +28,116 @@ Here are questions you can answer before you file a bug to make sure you're not
...
 
-
 ## Contributing code and content
+
 You will need to complete a Contributor License Agreement (CLA) before your pull request can be accepted. This agreement testifies that you are granting us permission to use the source code you are submitting, and that this work is being submitted under appropriate license that we can use it.
 
 You can complete the CLA by going through the steps at the [Contribution License Agreement site](https://cla.microsoft.com). Once we have received the signed CLA, we'll review the request. You will only need to do this once.
 
-Make sure you can build the code. Familiarize yourself with the project workflow and our coding conventions. If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. The hcttest command is your friend.
+Make sure you can build the code. Familiarize yourself with the project workflow and our coding conventions. If you don't know what a pull request is read this article: <https://help.github.com/articles/using-pull-requests>. The hcttest command is your friend.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:37]:
```suggestion
Make sure you can build the code. Familiarize yourself with the project workflow and our coding conventions. If you don't know what a pull request is read this article: <https://help.github.com/articles/using-pull-requests>.
```

Just noticed this. Since I think we document that as the "legacy" workflow we should probably not refer to it here.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -28,65 +28,116 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+Tags aren't formalized or any specific limited set.
+If you're unsure of a reasonable tag to use, just don't use any.
+If you want to invent a new tag, go for it!
+These are to help categorize changes at a glance.
+
+#### Descriptions
+
+The PR description should include a more detailed description of the change,
+ an explanation for the motivation of the change, and links to any relevant Issues.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:87]: nit: odd line wrapping.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -28,65 +28,116 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+This does not need to be a dissertation, but should leave breadcrumbs for the next person debugging your code (who might be you).
+
+Using the words `Fixes`, `Fixed`, `Closes`, `Closed`, or `Close` followed by
+ `#<issuenumber>`, will auto close an issue after the PR is merged.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:91]: nit: odd line wrap here too.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -28,65 +28,116 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+
+#### Release Notes
+
+Significant changes may require release notes that highlight important
+ compiler behavior changes for each named release.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:96]: nit: and here

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -28,65 +28,116 @@ Here are questions you can answer before you file a bug to make sure you're not
...
+- Links to specific PRs shouldn't be included.
+- Markdown links to bugs are encouraged if the issue is too complicated to completely explain in a single sentence.
+- Remember to update release notes as the nature of the change alters or is removed.
+
+### Testing Pull Requests
+
+All changes that make functional or behavioral changes to the compiler whether by fixing bugs or adding features
+ must include additional testing in the implementing pull request.
+Changes that do not change behavior may still be required to add testing if the change impacts areas with limited test coverage
+ to verify that the change doesn't alter previously untested, but important behavior.
```
Comment (llvm-beanz) [inline CONTRIBUTING.md:126]: nit: couple more odd line wraps here.

**Review by llvm-beanz — APPROVED** [review]
Some nits, but looks good.

File: CONTRIBUTING.md (hunk 14 above)
Comment (pow2clk) [reply CONTRIBUTING.md:126]: These are all intended. I find shorter lines easier to review and the space on the next line is meant to indicate a continuation of the previous line. When it's processed, both the newline and the space get converted into a single space. 

It's not a hill I'm determined to die on, but I think it's helpful for reviewing the source.

Comment (llvm-beanz) [reply CONTRIBUTING.md:126]: I think if you want line wrapping it should be consistent. This is crazy inconsistent across the whole file.
